flink-tests/src/test/java/org/apache/flink/test/scheduling/ReactiveModeITCase.java

```diff
@@ -0,0 +1,239 @@
+        ParallelismTrackingSink.waitForInstances();
+    }
+
+    private int getNumberOfConnectedTaskManagers() throws ExecutionException, InterruptedException {
+        return miniClusterResource
+                .getMiniCluster()
+                .requestClusterOverview()
+                .get()
+                .getNumTaskManagersConnected();
+    }
+
+    private void startAdditionalTaskManager() throws Exception {
+        miniClusterResource.getMiniCluster().startTaskManager();
+        CommonTestUtils.waitUntilCondition(
+                () -> getNumberOfConnectedTaskManagers() == 2,
+                Deadline.fromNow(Duration.ofMillis(10_000L)));
+    }
+
+    private static class ParallelismTrackingSource implements ParallelSourceFunction<String> {
+        private volatile boolean running = true;
+
+        private static final InstanceTracker instances = new InstanceTracker();
+
+        public static void expectInstances(int count) {
+            instances.expectInstances(count);
+        }
+
+        public static void waitForInstances() throws InterruptedException {
+            instances.waitForInstances();
+        }
+
+        @Override
+        public void run(SourceContext<String> ctx) throws Exception {
+            instances.reportNewInstance();
+            while (running) {
+                synchronized (ctx.getCheckpointLock()) {
+                    ctx.collect("test");
+                }
+                Thread.sleep(100);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            running = false;
+        }
+    }
+
+    private static class ParallelismTrackingSink<T> extends RichSinkFunction<T> {
+
+        private static final InstanceTracker instances = new InstanceTracker();
+
+        public static void expectInstances(int count) {
+            instances.expectInstances(count);
+        }
+
+        public static void waitForInstances() throws InterruptedException {
+            instances.waitForInstances();
+        }
+
+        @Override
+        public void open(Configuration parameters) throws Exception {
+            instances.reportNewInstance();
+        }
+    }
+
+    private static class InstanceTracker {
+        private final Object lock = new Object();
+
+        @GuardedBy("lock")
+        private CountDownLatch latch = new CountDownLatch(0);
+
+        public void reportNewInstance() {
+            synchronized (lock) {
+                if (latch.getCount() == 0) {
+                    throw new RuntimeException("Test error. More instances than expected.");
+                }
+                latch.countDown();
+            }
+        }
+
+        public void waitForInstances() throws InterruptedException {
+            while (true) {
+                synchronized (lock) {
+                    if (latch.await(10, TimeUnit.MILLISECONDS)) {
+                        break;
+                    }
+                }
+            }
```

Comment:
Why not simply `synchronized(lock) { latch.await(); }`?

Reply:
We need to release the lock from time to time to allow other threads acquiring it. In this particular case, if we are waiting for the latch to complete while holding the lock, we can never enter the lock in `reportNewInstance` to count down the latch.